backend/GoogleAPI.ts

```diff
@@ -7,33 +8,44 @@ const PATH_TO_KEY_FILE = process.env.GOOGLE_SERVICE_ACCOUNT_JSON_FILE;
 
 const SPREADSHEET_ID = process.env.SPREADSHEET_ID;
 
-const AUTHENTICATION = new google.auth.GoogleAuth({
-  keyFile: PATH_TO_KEY_FILE,
-  scopes: SCOPES,
-});
-
-const SHEETS = google.sheets({
-  version: 'v4',
-  auth: AUTHENTICATION,
-});
-
 export async function sendFeedbackToGoogleSheets({
   feedbackText,
   userDetails,
 }: {
   feedbackText: string;
   userDetails: User;
 }) {
-  const date = `${new Date().toLocaleDateString('fi-FI')} ${new Date().toTimeString()}`;
-  const feedbackGiverDetails = `${userDetails.firstName} ${userDetails.lastName}\n${userDetails.email}`;
-  await SHEETS.spreadsheets.values.append({
-    auth: AUTHENTICATION,
-    spreadsheetId: SPREADSHEET_ID,
-    range: `Palautteet!A:C`,
-    requestBody: {
-      majorDimension: 'ROWS',
-      values: [[feedbackText, feedbackGiverDetails, date]],
-    },
-    valueInputOption: 'USER_ENTERED',
-  });
+  try {
+    const AUTHENTICATION = process.env.GOOGLE_SERVICE_ACCOUNT_CREDENTIALS
+      ? new google.auth.GoogleAuth({
+          credentials: JSON.parse(
+            process.env.GOOGLE_SERVICE_ACCOUNT_CREDENTIALS,
+          ) as JWTInput,
+          scopes: SCOPES,
+        })
+      : new google.auth.GoogleAuth({
+          keyFile: PATH_TO_KEY_FILE,
+          scopes: SCOPES,
+        });
```

Comment:
Lisäsin tähän hätään tsekkauksen, että jos tuo variable on täytty `.env`-tiedostossa, käytetään sitä. Käytin tässä `as`, koska en keksinyt helppoa tapaa tsekata tuon typeä. `JSON.parse` palauttaa `any`:n. Zodille ei taida pystyä laittamaan valmiiksi olevaa typeä :thinking:

Reply:
Siirtäisin credentiaalit `GoogleAuth` funktion ulkopuolelle. `as` toimii tässä, jos haluaa käyttää zodia, niin pitää itse tehdä skeema sitä varten.
```suggestion
    const credentialsOrKeyFile = process.env.GOOGLE_SERVICE_ACCOUNT_CREDENTIALS
      ? {
          credentials: JSON.parse(
            process.env.GOOGLE_SERVICE_ACCOUNT_CREDENTIALS,
          ) as JWTInput,
        }
      : { keyFile: PATH_TO_KEY_FILE };

    const auth = new google.auth.GoogleAuth({
      ...credentialsOrKeyFile,
      scopes: SCOPES,
    });
```

Reply:
Tämä oli hyvä idea, paljon parempi! Kiitos! :+1: